Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Data/ExportImport2/CatalogFullExportPagedDataSource.cs

```diff
@@ -0,0 +1,182 @@
+                    FetchFunc = async (x) =>
+                    {
+                        var searchResult = await _categorySearchService.SearchCategoriesAsync((CategorySearchCriteria)x.SearchCriteria);
+                        x.TotalCount = searchResult.TotalCount;
+                        x.Result = searchResult.Results;
+                    }
+                });
+
+                _exportDataSourceStates.Add(new ExportDataSourceState
+                {
+                    SearchCriteria = new PropertySearchCriteria { CatalogIds = _dataQuery.CatalogIds },
+                    FetchFunc = async (x) =>
+                    {
+                        var searchResult = await _propertySearchService.SearchPropertiesAsync((PropertySearchCriteria)x.SearchCriteria);
+                        x.TotalCount = searchResult.TotalCount;
+                        x.Result = searchResult.Results;
+                    }
+                });
+
+                _exportDataSourceStates.Add(new ExportDataSourceState
+                {
+                    SearchCriteria = new PropertyDictionaryItemSearchCriteria() { CatalogIds = _dataQuery.CatalogIds },
+                    FetchFunc = async (x) =>
+                    {
+                        var searchResult = await _propertyDictionarySearchService.SearchAsync((PropertyDictionaryItemSearchCriteria)x.SearchCriteria);
+                        x.TotalCount = searchResult.TotalCount;
+                        x.Result = searchResult.Results;
+                    }
+                });
+
+                _exportDataSourceStates.Add(new ExportDataSourceState
+                {
+                    SearchCriteria = new ProductSearchCriteria() { CatalogIds = _dataQuery.CatalogIds, ResponseGroup = (ItemResponseGroup.Full & ~ItemResponseGroup.Variations).ToString() },
+                    FetchFunc = async (x) =>
+                    {
+                        var searchResult = await _productSearchService.SearchProductsAsync((ProductSearchCriteria)x.SearchCriteria);
+                        x.TotalCount = searchResult.TotalCount;
+                        x.Result = searchResult.Results;
+                    }
+                });
+
+                CalculateCounts();
+            }
+        }
+
+        public int GetTotalCount()
+        {
+            CalculateCounts();
+            return TotalCount;
+        }
+
+        private void EnsureHaveTotals()
+        {
+            if (TotalCount < 0)
+            {
+                CalculateCounts();
+            }
+        }
+
+        public bool Fetch()
+        {
+            EnsureHaveTotals();
+
+            var take = _dataQuery.Take ?? PageSize;
+            var skip = _dataQuery.Skip ?? CurrentPageNumber * PageSize;
+            var taskList = new List<Task>();
+            Items = new List<IExportable>();
+
+            var curStateStartPos = 0; // Starting position of current state
```

Comment:
If name the variable properly there is no need in such comments.

Reply:
Also I think we even do not need current position in state. As we always have Skip and Take, and state TotalCount's. These data are sufficient to take portion of data.